hera_pspec/grouping.py

```diff
@@ -288,74 +307,135 @@ def average_spectra(uvp_in, blpair_groups=None, time_avg=False,
-
-                      for stat in stat_l:
-                        errws[stat] = np.sum(errws[stat], axis=0)[None]
-                    # Add multiple copies of data for each baseline according
-                    # to the weighting/multiplicity
-                    for m in range(int(blpg_wgts[k])):
-                        bpg_data.append(data * w)
-                        bpg_wgts.append(wgts * w[:, None])
-                        bpg_ints.append(ints * w)
-                        bpg_nsmp.append(nsmp)
-                        [bpg_stats[stat].append(errws[stat]) for stat in stat_l]
-                        w_list.append(w)
-                        if store_cov:
-                            bpg_cov.append(cov * w[:, None])
 
+                    if if_error_weighted:
+                    # If if_error_weighted==True, all arrays are weighted by a specified kind of errors.
+                    # For each power spectrum P_i with error sigma_i, 
+                    # P_avg = \sum{ P_i / (sigma_i)^2 } / \sum{ 1 / (sigma_i)^2 } 
+                    # while for terms like variance or covariance,
+                    # epsilon_avg = \sum{ (epsilon_i / (sigma_i)^4 } / ( \sum{ 1 / (sigma_i)^2 } )^2
+                    # For reference: Tegmark 1997, The Astrophysical Journal Letters, 480, L87, Table 1, #3
+                    # or Dillon 2014, Physical Review D, 89, 023002 , Equation 34. 
+                        error_weights = uvp.get_stats(name_error_weights, (spw, blp, p)).copy()
+                        error_weights = 1./ error_weights**2
+                        # shape of error_weights: (Ntimes, Ndlys)
+                        
+                        # Get error bar
+                        errws = {}
+                        for stat in stat_l:
+                            errs = uvp.get_stats(stat, (spw, blp, p)).copy()
+                            # shape of errs: (Ntimes, Ndlys)
+                            errws[stat] = errs**2
+
+                        # Take time average if desired
+                        if time_avg:
+                            data = (np.sum(data * error_weights, axis=0) \
+                                / np.sum(error_weights, axis=0))[None]
+                            wgts = (np.sum(wgts * error_weights[:, :, None], axis=0) \
+                                / np.sum(error_weights, axis=0)[:, None])[None]
+                            ints = (np.sum(ints * error_weights, axis=0) \
+                                / np.sum(error_weights, axis=0))[None]
+                            nsmp = np.sum(nsmp, axis=0)[None]
+                            if store_cov:
+                                cov = (np.sum(cov * error_weights[:, :, None]**2, axis=0) \
+                                    / (np.sum(w,axis=0))[:, None]**2)[None]
+                            for stat in stat_l:
+                                errws[stat] = (np.sum(errws[stat]*error_weights**2, axis=0) \
```

Comment:
I think this is doing a weighted sum (in quadrature) of all of the error fields, where the weights are from the field specified by `name_error_weights` rather than each error field itself. Perhaps this should be mentioned in the docstring at the top, as it's slightly non-trivial.

Reply:
Sure.